src/std/error.ss

```diff
@@ -1,41 +1,469 @@
    runtime-identifier: (quote-syntax error::t)))
 
-(defmethod {display-exception <error>}
+;; Error base class
+(defclass (Error <error>) ()
+  constructor: :init!)
+
+(defmethod {:init! Error}
+  (lambda (self message where: (where #f) irritants: (irritants []))
+    (struct-instance-init! self message irritants where)))
+
+(defmethod {display-exception Error}
   (lambda (self port)
-    (parameterize ((current-output-port port))
-      (cond
-       ((error-trace self)
-        => (lambda (where) (display* where ": "))))
-      (display* "[" (##type-name (object-type self)) "] ")
-      (displayln (error-message self))
-      (let (irritants (error-irritants self))
-        (unless (null? irritants)
-          (display "--- irritants: ")
-          (for-each
-            (lambda (obj) (display* obj " "))
-            irritants)
-          (newline)))))
+    (let (old-width (fix-port-width! port))
+      (parameterize ((current-output-port port))
+        (cond
+         ((Error-where self)
+          => (lambda (where) (display* where ": "))))
+        (display* "[" (##type-name (object-type self)) "] ")
+        (displayln (Error-message self))
+        (let (irritants (Error-irritants self))
+          (unless (null? irritants)
+            (display "--- irritants: ")
+            (for-each
+              (lambda (obj) (display* obj " "))
+              irritants)
+            (newline)))
+        (when (StackTrace? self)
+          (alet (cont (StackTrace-continuation self))
+            (displayln "--- continuation backtrace:")
+            (display-continuation-backtrace cont))))
+      (reset-port-width! port old-width)))
   rebind: #t)
 
-(defstruct (io-error <error>) ())
-(defstruct (timeout-error <error>) ())
+(def (Error-message err)
+  (##structure-ref err 1 error::t #f))
+(def (Error-irritants err)
+  (##structure-ref err 2 error::t #f))
+(def (Error-where err)
+  (##structure-ref err 3 error::t #f))
+
+;; utility macro
+(defsyntax (deferror-class stx)
+  (syntax-case stx ()
+    ((_ Class slots predicate-alias)
+     (identifier? #'Class)
+     #'(deferror-class (Class StackTrace Error) slots predicate-alias Error:::init!))
+    ((_ Class slots predicate-alias kons)
+     (identifier? #'Class)
+     #'(deferror-class (Class StackTrace Error) slots predicate-alias kons))
+    ((_ (Class Mixin ...) slots predicate-alias)
+     (identifier-list? #'(Class Mixin ...))
+     #'(deferror-class (Class Mixin ...) slots predicate-alias Error:::init!))
+    ((_ (Class Mixin ...) slots predicate-alias kons)
+     (identifier-list? #'(Class Mixin ...))
+     (with-syntax ((Class? (stx-identifier #'Class #'Class "?")))
+       #'(begin
+           (defclass (Class Mixin ...) slots)
+           (defmethod {:init! Class}
+             kons)
+           (def predicate-alias Class?))))))
+
+;; Mixin for getting stack traces
+(defclass StackTrace (continuation))
+
+;; Input argument errors
+(deferror-class BadArgument () bad-argument-error?)
```

Comment:
Why not BadArgument?

Reply:
Or else bad-argument-error, or bad-argument. I don't like this naming non-convention.

Reply:
BadArgumet? Is there, the pred is an alias for schemier names to match the raisers.